src/coreclr/nativeaot/Runtime/arm/Interlocked.S

```diff
@@ -73,30 +61,3 @@ LOCAL_LABEL(CmpXchg32Exit):
           dmb
           bx           lr
 LEAF_END RhpLockCmpXchg32, _TEXT
-
-// WARNING: Code in EHHelpers.cpp makes assumptions about this helper, in particular:
-// - Function "InWriteBarrierHelper" assumes an AV due to passed in null pointer will happen at RhpLockCmpXchg64AVLocation
-// - Function "UnwindSimpleHelperToCaller" assumes no registers were pushed and LR contains the return address
-// r0      = destination address
-// {r2,r3} = value
-// sp[0+8] = comparand
-LEAF_ENTRY RhpLockCmpXchg64, _TEXT
-GLOBAL_LABEL RhpLockCmpXchg64AVLocation
-          ldr          r12, [r0]        // dummy read for null check
-          PROLOG_PUSH  "{r4-r6,lr}"
-          dmb
-          ldrd         r4, r5, [sp,#0x10]
-LOCAL_LABEL(CmpXchg64Retry):
-          ldrexd       r6, r1, [r0]
-          cmp          r6, r4
-          bne          LOCAL_LABEL(CmpXchg64Exit)
-          cmp          r1, r5
-          bne          LOCAL_LABEL(CmpXchg64Exit)
-          strexd       r12, r2, r3, [r0]
-          cmp          r12, #0
-          bne          LOCAL_LABEL(CmpXchg64Retry)
-LOCAL_LABEL(CmpXchg64Exit):
-          mov          r0, r6
-          dmb
```

Comment:
The native AOT PAL implementation of `InterlockedCompareExchange64` seems to be missing this barrier (`PAL_InterlockedOperationBarrier` in CoreCLR PAL).

Reply:
I added it but I noticed that `PAL_InterlockedOperationBarrier` is not enabled for arm32. That seems odd.

Reply:
> I added it but I noticed that `PAL_InterlockedOperationBarrier` is not enabled for arm32. That seems odd.

IIRC the native compilers already emit a barrier of their own there so it'd do two barriers (so I guess it's kinda relying on an implementation detail).

Reply:
> the native compilers already emit a barrier of their own

At very least ARM gcc and ARM clang does, so that explains why it wasn't a problem for ARM32 on CoreCLR.

Reply:
ARM64: clang generates barrier at the end (should still be fine); gcc doesn't generate a barrier